File: apps/api/v1/core/dues.py
"""Monthly dues calculation — mirrors packages/utils/src/dues.ts"""

from datetime import date, datetime, timezone

MONTHLY_DUES_AMOUNT = 30.0
DUES_CURRENCY = "GHS"
DUES_EFFECTIVE_FROM = (2025, 1)
DUES_REMINDER_DAYS_BEFORE_MONTH_END = 3

MONTH_NAMES = [
    "January",
    "February",
    "March",
    "April",
    "May",
    "June",
    "July",
    "August",
    "September",
    "October",
    "November",
    "December",
]


def period_label(year: int, month: int) -> str:
    return f"{MONTH_NAMES[month - 1]} {year}"


def _period_key(year: int, month: int) -> str:
    return f"{year}-{month:02d}"


def _compare_period(a: tuple[int, int], b: tuple[int, int]) -> int:
    if a[0] != b[0]:
        return a[0] - b[0]
    return a[1] - b[1]


def _add_months(year: int, month: int, delta: int) -> tuple[int, int]:
    d = date(year, month, 1)
    m = month - 1 + delta
    y = year + m // 12
    m = m % 12 + 1
    return y, m


def _iter_due_periods(member_since: tuple[int, int], as_of: tuple[int, int]) -> list[tuple[int, int]]:
    start = member_since if _compare_period(member_since, DUES_EFFECTIVE_FROM) > 0 else DUES_EFFECTIVE_FROM
    periods: list[tuple[int, int]] = []
    cur = start
    while _compare_period(cur, as_of) <= 0:
        periods.append(cur)
        cur = _add_months(cur[0], cur[1], 1)
    return periods


def compute_dues_summary(
    member_since: date | datetime,
    paid_periods: list[dict],
    as_of: datetime | None = None,
    monthly_amount: float | None = None,
) -> dict:
    if as_of is None:
        as_of = datetime.now(timezone.utc)

    monthly = float(monthly_amount if monthly_amount is not None else MONTHLY_DUES_AMOUNT)

    if isinstance(member_since, datetime):
        member_since = member_since.date()

    member_start = (member_since.year, member_since.month)
    current = (as_of.year, as_of.month)

    paid_map: dict[str, float] = {}
    for p in paid_periods:
        year = int(p["year"])
        month = int(p["month"])
        key = _period_key(year, month)
        paid_map[key] = paid_map.get(key, 0.0) + float(p["amount"])

    due_periods = _iter_due_periods(member_start, current)

    arrears_count = 0
    total_owed = 0.0
    total_paid_months = 0
    periods: list[dict] = []

    for year, month in due_periods:
        paid_amount = paid_map.get(_period_key(year, month), 0.0)
        is_paid = paid_amount >= monthly
        is_current = year == current[0] and month == current[1]
        is_future = _compare_period((year, month), current) > 0
        is_past = _compare_period((year, month), current) < 0

        if is_paid:
            status = "paid"
            total_paid_months += 1
        elif is_future:
            status = "upcoming"
        elif is_current:
            status = "due"
        else:
            status = "overdue"
            if is_past:
                arrears_count += 1
                total_owed += monthly - paid_amount

        periods.append(
            {
                "year": year,
                "month": month,
                "label": period_label(year, month),
                "amount": monthly,
                "status": status,
                "paid_amount": round(paid_amount, 2),
            }
        )

    current_period = next((p for p in periods if p["year"] == current[0] and p["month"] == current[1]), None)
    current_status = "due"
    if current_period:
        if current_period["status"] == "paid":
            current_status = "paid"
        elif current_period["status"] == "overdue":
            current_status = "overdue"

    return {
        "monthly_amount": monthly,
        "currency": DUES_CURRENCY,
        "current_month": {
            "year": current[0],
            "month": current[1],
            "label": period_label(current[0], current[1]),
        },
        "current_status": current_status,
        "arrears_count": arrears_count,
        "total_owed": round(total_owed, 2),
        "total_paid_months": total_paid_months,
        "periods": periods,
    }

File: packages/utils/src/dues.ts
/** Mirrors packages/config/src/dues.ts — keep values in sync */

const DUES = {
  MONTHLY_AMOUNT: 30,
  CURRENCY: "GHS",
  EFFECTIVE_FROM: { year: 2025, month: 1 },
} as const;

export type DuesPeriodStatus = "paid" | "due" | "overdue" | "upcoming";

export interface DuesPeriod {
  year: number;
  month: number;
  label: string;
  amount: number;
  status: DuesPeriodStatus;
  paidAmount: number;
}

export interface DuesSummary {
  monthlyAmount: number;
  currency: string;
  currentMonth: { year: number; month: number; label: string };
  currentStatus: "paid" | "due" | "overdue";
  arrearsCount: number;
  totalOwed: number;
  totalPaidMonths: number;
  periods: DuesPeriod[];
}

export interface PaidPeriodInput {
  year: number;
  month: number;
  amount: number;
}

const MONTH_NAMES = [
  "January",
  "February",
  "March",
  "April",
  "May",
  "June",
  "July",
  "August",
  "September",
  "October",
  "November",
  "December",
] as const;

export function periodLabel(year: number, month: number): string {
  return `${MONTH_NAMES[month - 1]} ${year}`;
}

function periodKey(year: number, month: number): string {
  return `${year}-${String(month).padStart(2, "0")}`;
}

function comparePeriod(a: { year: number; month: number }, b: { year: number; month: number }): number {
  if (a.year !== b.year) return a.year - b.year;
  return a.month - b.month;
}

function addMonths(year: number, month: number, delta: number): { year: number; month: number } {
  const d = new Date(year, month - 1 + delta, 1);
  return { year: d.getFullYear(), month: d.getMonth() + 1 };
}

function iterDuePeriods(
  memberSince: { year: number; month: number },
  asOf: { year: number; month: number }
): Array<{ year: number; month: number }> {
  const start = comparePeriod(memberSince, DUES.EFFECTIVE_FROM) > 0 ? memberSince : DUES.EFFECTIVE_FROM;
  const periods: Array<{ year: number; month: number }> = [];
  let cur = { ...start };
  while (comparePeriod(cur, asOf) <= 0) {
    periods.push({ ...cur });
    cur = addMonths(cur.year, cur.month, 1);
  }
  return periods;
}

export function computeDuesSummary(
  memberSince: Date | string,
  paidPeriods: PaidPeriodInput[],
  asOf: Date = new Date(),
  monthlyAmount: number = DUES.MONTHLY_AMOUNT
): DuesSummary {
  const since = typeof memberSince === "string" ? new Date(memberSince) : memberSince;
  const memberStart = { year: since.getFullYear(), month: since.getMonth() + 1 };
  const current = { year: asOf.getFullYear(), month: asOf.getMonth() + 1 };

  const paidMap = new Map<string, number>();
  for (const p of paidPeriods) {
    const key = periodKey(p.year, p.month);
    paidMap.set(key, (paidMap.get(key) ?? 0) + p.amount);
  }

  const duePeriods = iterDuePeriods(memberStart, current);

  let arrearsCount = 0;
  let totalOwed = 0;
  let totalPaidMonths = 0;

  const periods: DuesPeriod[] = duePeriods.map(({ year, month }) => {
    const paidAmount = paidMap.get(periodKey(year, month)) ?? 0;
    const isPaid = paidAmount >= monthlyAmount;
    const isCurrent = year === current.year && month === current.month;
    const isFuture = comparePeriod({ year, month }, current) > 0;
    const isPast = comparePeriod({ year, month }, current) < 0;

    let status: DuesPeriodStatus;
    if (isPaid) {
      status = "paid";
      totalPaidMonths += 1;
    } else if (isFuture) {
      status = "upcoming";
    } else if (isCurrent) {
      status = "due";
    } else {
      status = "overdue";
      if (isPast) {
        arrearsCount += 1;
        totalOwed += monthlyAmount - paidAmount;
      }
    }

    return {
      year,
      month,
      label: periodLabel(year, month),
      amount: monthlyAmount,
      status,
      paidAmount,
    };
  });

  const currentPeriod = periods.find((p) => p.year === current.year && p.month === current.month);
  let currentStatus: "paid" | "due" | "overdue" = "due";
  if (currentPeriod?.status === "paid") currentStatus = "paid";
  else if (currentPeriod?.status === "overdue") currentStatus = "overdue";

  return {
    monthlyAmount,
    currency: DUES.CURRENCY,
    currentMonth: { ...current, label: periodLabel(current.year, current.month) },
    currentStatus,
    arrearsCount,
    totalOwed,
    totalPaidMonths,
    periods,
  };
}
